scheduler/src/cook/mesos/scheduler.clj

```diff
@@ -436,112 +445,232 @@
 
-(defn handle-resource-offer!
+(defn job-allowed-to-start?
+  "Converts the DB function :job/allowed-to-start? into a predicate"
+  [db job]
+  (try
+    (d/invoke db :job/allowed-to-start? db (:db/id job))
+    true
+    (catch clojure.lang.ExceptionInfo e
+      false)))
+
+(defn ids-of-backfilled-instances
+  "Returns a list of the ids of running, backfilled instances of the given job"
+  [job]
+  (->> (:job/instance job)
+       (filter #(and (= :instance.status/running (:instance/status %))
+                     (:instance/backfilled? %)))
+       (map :db/id)))
+
+(defn process-matches-for-backfill
+  "This computes some sets:
+   
+   fully-processed: this is a set of job uuids that should be removed from the list scheduler jobs. These are not backfilled
+   upgrade-backfill: this is a set of instance datomic ids that should be upgraded to non-backfill
+   backfill-jobs: this is a set of job uuids for jobs that were matched, but are in the backfill QoS class
+   "
+  [scheduler-contents head-of-considerable matched-jobs]
+  (let [matched-job-uuids (set (mapv :job/uuid matched-jobs))
+        head-uuid (:job/uuid head-of-considerable)]
+    (loop [scheduler-contents scheduler-contents
+           backfilling? false
+           matched-head? false
+           fully-processed #{}
+           upgrade-backfill #{}
+           backfill-jobs #{}]
+      (if (seq scheduler-contents)
+        (let [scheduler-contents' (rest scheduler-contents)
+              pending-job (first scheduler-contents)
+              pending-job-uuid (:job/uuid pending-job)
+              pending=head? (= pending-job-uuid head-uuid)]
+          (if (contains? matched-job-uuids pending-job-uuid)
+            (if backfilling?
+              (recur scheduler-contents' true matched-head? fully-processed upgrade-backfill (conj backfill-jobs pending-job-uuid))
+              (recur scheduler-contents' false (or pending=head? matched-head?) (conj fully-processed pending-job-uuid) upgrade-backfill backfill-jobs))
+            (if-let [backfilled-ids (and (not backfilling?) (seq (ids-of-backfilled-instances pending-job)))]
+              (recur scheduler-contents' false (or pending=head? matched-head?) fully-processed (into upgrade-backfill backfilled-ids) backfill-jobs)
+              (recur scheduler-contents' true matched-head? fully-processed upgrade-backfill backfill-jobs))))
+        {:fully-processed fully-processed
+         :upgrade-backfill upgrade-backfill
+         :backfill-jobs backfill-jobs
+         :matched-head? matched-head?}))))
+
+(defn handle-resource-offers!
   "Gets a list of offers from mesos. Decides what to do with them all--they should all
    be accepted or rejected at the end of the function."
-  [conn driver fid pending-jobs offer]
-  (try
+  [conn driver ^TaskScheduler fenzo fid pending-jobs num-considerable offers-chan offers]
+  (doseq [offer offers]
```

Comment:
I don't think this is the right place to do this. We will end up double counting in cases where there is an error and the offer stash is piped back in.

Move this expression do into `make-offer-handler`